src/admin/index.js

```diff
@@ -990,6 +994,86 @@ module.exports = ({
     })
   }
 
+  /**
+   * Delete topic records up to the selected partition offsets
+   *
+   * @param {string} topic
+   * @param {Array<SeekEntry>} partitions
+   * @return {Promise}
+   *
+   * @typedef {Object} SeekEntry
+   * @property {number} partition
+   * @property {string} offset
+   */
+  const deleteTopicRecords = async ({ topic, partitions }) => {
+    if (!topic || typeof topic !== 'string') {
+      throw new KafkaJSNonRetriableError(`Invalid topic ${topic}`)
+    }
+
+    if (!partitions || partitions.length === 0) {
+      throw new KafkaJSNonRetriableError(`Invalid partitions`)
+    }
+
+    const partitionsByBroker = cluster.findLeaderForPartitions(
+      topic,
+      partitions.map(p => p.partition)
+    )
+
+    const partitionsFound = flatten(values(partitionsByBroker))
+    const topicOffsets = await fetchTopicOffsets(topic)
+
+    partitions.forEach(({ partition, offset }) => {
+      // warn if no broker/leader found for partition
+      if (!partitionsFound.includes(partition)) {
+        logger.warn('Could not find broker/leader for the partition', {
+          topic,
+          partition,
+        })
+        return
+      }
+      const { high, low } = topicOffsets.find(p => p.partition === partition)
+      // throw in case of offset out-of-range (otherwise could fail mid-process)
+      if (parseInt(offset) > parseInt(high) || parseInt(offset) < -1) {
```

Comment:
Do we really need all these checks? AFAIK, they are already handled on the broker side. I'm not sure that I see a compelling enough reason to handle them on the client side when we already have to deal with partial failures regardless. The response should already contain all the information needed to know which partitions failed and for what reason.

Reply:
(resolving as this comment has been covered by the updated KafkaJSDeleteTopicRecordsError pattern, I think?)